utilities/src/with_std/dynamic_events.rs

```diff
@@ -0,0 +1,419 @@
+
+		let pallet_index = dispatch_error
+			.pointer("/Module/index")?
+			.as_u64()
+			.expect("/Module/index must be a number that fits in a u8") as u8;
+
+		log::debug!("Pallet index: {}", pallet_index);
+
+		self.errors.get(&pallet_index).and_then(|(pallet, pallet_errors)| {
+			pallet_errors.get(&error_index).map(|(name, doc)| {
+				JsonValue::Object(
+					[("pallet", pallet.clone()), ("name", name.clone()), ("error", doc.join(" "))]
+						.into_iter()
+						.map(|(k, v)| (k.to_string(), JsonValue::String(v)))
+						.collect(),
+				)
+			})
+		})
+	}
+}
+
+#[cfg(test)]
+mod test {
+	use super::*;
+
+	#[test]
+	fn test_event_wrappers() {
+		let invalid = serde_json::json!({
+			"phase": {
+				"ApplyExtrinsic": 401
+			},
+			"topics": "0x"
+		});
+		assert!(DynamicEventRecord::new(invalid).is_none());
+
+		let success = serde_json::json!({
+			"event": {
+				"System": {
+					"ExtrinsicSuccess": {
+						"dispatch_info": {
+							"class": "Normal",
+							"pays_fee": "Yes",
+							"weight": {
+								"proof_size": "0x1efd",
+								"ref_time": "0x282deb50"
+							}
+						}
+					},
+				}
+			},
+			"phase": {
+				"ApplyExtrinsic": 401
+			},
+			"topics": "0x"
+		});
+		let failed_1 = serde_json::json!({
+			"event": {
+				"System": {
+					"ExtrinsicFailed": {
+						"dispatch_error": {
+							"Module": {
+								"error": "0x00000000",
+								"index": 31
+							}
+						},
+						"dispatch_info": {
+							"class": "Normal",
+							"pays_fee": "Yes",
+							"weight": {
+								"proof_size": "0x1efd",
+								"ref_time": "0x282deb50"
+							}
+						}
+					}
+				}
+			},
+			"phase": {
+				"ApplyExtrinsic": 401
+			},
+			"topics": "0x"
+		});
+		let failed_2 = serde_json::json!({
+			"event": {
+				"System": {
+					"ExtrinsicFailed": {
+						"dispatch_error":{
+							"error": "The user does not have enough funds.",
+							"name": "InsufficientBalance",
+							"pallet": "LiquidityProvider"
+						},
+						"dispatch_info": {
+							"class": "Normal",
+							"pays_fee": "Yes",
+							"weight": {
+								"proof_size": "0x1efd",
+								"ref_time": "0x282deb50"
+							}
+						}
+					}
+				}
+			},
+			"phase": {
+				"ApplyExtrinsic": 401
+			},
+			"topics": "0x"
+		});
+
+		for json in [&success, &failed_1, &failed_2] {
+			let record = DynamicEventRecord::new(json.clone()).unwrap();
+			let (index, event) = record.indexed_extrinsic_event().unwrap();
+			assert_eq!(index, 401);
+			assert_eq!(record.extrinsic_event(index).as_deref(), Some(event).as_deref());
+		}
+
+		let outcome = |json| {
+			DynamicEventRecord::new(json)
+				.expect("Valid json event record")
+				.event()
+				.extrinsic_outcome()
+				.expect("outcome should not be None")
+		};
+
+		assert!(outcome(success).is_ok());
+		assert!(matches!(outcome(failed_1).unwrap_err(), DispatchError::DispatchError(_)));
+		assert!(matches!(outcome(failed_2).unwrap_err(), DispatchError::KnownModuleError { .. }));
```

Comment:
think we also check the module is correct?

Reply:
This isn't testing any decoding, it's really just passing through the json strings verbatim, so there's not much value in testing this... but I suppose sure, why not.

I agree there should be a test for the error decoding though. 

FWIW the error conversion code was mostly lifted from the scc, I can pull across the tests from there (and maybe add a few extra, we'll see).